Add support for InterpolatedStringBuilderArgumentAttribute

333fred opened the pull request:
This adds binding and lowering support for InterpolatedStringBuilderArgumentAttribute. For initial binding, we have an existing method that all argument binding goes through: CoerceArguments. However, lowering did not have a similar chokepoint, as every argument location started by calling `VisitList`, then `MakeArguments` at a later point. In order to ensure that the rewrite is handled correctly, I've done a refactor on `MakeArguments` to turn it into `VisitArguments` and update every place that handles arguments to instead use this new helper, which takes care of the rest of the argument rewrites (such as params) and will additionally update the receiver if needed.

Test plan: https://github.com/dotnet/roslyn/issues/51499

333fred commented:
@chsienki @AlekseyTs for review.

333fred commented on src/Compilers/CSharp/Test/Symbol/Symbols/IndexedPropertyTests.cs:
```diff
@@ -412,22 +412,22 @@ .maxstack  3
   .locals init (IA V_0,
-  int V_1,
-  int V_2)
+                int V_1,
```
The changes in this file are just slot order changes: the resulting effect is unchanged, we just use the earlier slot before the later slot now.

chsienki commented on src/Compilers/Core/Portable/InternalUtilities/SpanUtilities.cs:
```diff
@@ -20,5 +20,8 @@ internal static class SpanUtilities
         }
+
+        public static bool All<TElement>(this ReadOnlySpan<TElement> span, Func<TElement, bool> predicate)
+            => span.All(predicate, static (item, predicate) => predicate(item));
```
I assume this isn't involved in any particuarly hot paths? Otherwiwse consider just duplicating the method to avoid the double function call.

AlekseyTs replied:
>=> span.All(predicate, static (item, predicate) => predicate(item)); [](http://example.com/codeflow?start=12&length=68)

I think we should just clone the body of the method above and make necessary adjustments instead of creating delegates that call other delegates.

chsienki commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -5216,6 +5241,7 @@ private bool IsConstructorAccessible(MethodSymbol constructor, ref CompoundUseSi
             NamedTypeSymbol type,
             AnalyzedArguments analyzedArguments,
             BindingDiagnosticBag diagnostics,
+            out bool overloadResolutionSucceeded,
```
Trying to understand this parameter a little bit. Looks like you're using in the string binding to say 'if overload resolution failed' then try again with a different set of params. In that case wont you get a `BoundBadExpression` with `LookupResultKind.OverloadResolutionFailure`?

333fred replied:
This is a good suggestion, I think that should totally work.

chsienki commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -4406,24 +4427,28 @@ private BoundExpression BindClassCreationExpression(ObjectCreationExpressionSynt
             SyntaxNode node,
-            BindingDiagnosticBag diagnostics)
+            BindingDiagnosticBag diagnostics,
+            out bool overloadResolutionSucceeded)
```
As below, is this required or is the `LookupResultKind` on the returned expression enough to decide?

chsienki commented on src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:
```diff
@@ -1009,13 +1009,7 @@ private void CheckRestrictedTypeReceiver(BoundExpression expression, CSharpCompi
 
             var methodResult = result.ValidResult;
             var returnType = methodResult.Member.ReturnType;
-            this.CoerceArguments(methodResult, analyzedArguments.Arguments, diagnostics);
```
Any reason for re-arranging these lines?

333fred replied:
Yes: `CoerceArguments` needs the receiver info.

chsienki commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -271,26 +280,119 @@ private BoundInterpolatedString BindUnconvertedInterpolatedStringToHandlerType(B
+                    // for a different reason, that different reason is the one the user will want to know about. If both or neither failed
+                    // because of this error, we'll report everything.
+
+                    var nonOutConstructorHasArityError = nonOutConstructorDiagnostics.DiagnosticBag?.AsEnumerableWithoutResolution().Any(d => (ErrorCode)d.Code == ErrorCode.ERR_BadCtorArgCount) ?? false;
```
It feels weird to inspect the diagnostics to figure out what happened. Can we expose this logic from the overload resolution result and report it from `TryPerformConstructorOverloadResolution` ?

333fred replied:
No, that doesn't work here. For example, if there's a `ref` mismatch on one both results will get a resultkind of `OverloadResolutionFailure`, but the refkind mismatch error is the better one to report.

chsienki commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_AssignmentOperator.cs:
```diff
@@ -309,17 +312,16 @@ private BoundExpression VisitAssignmentOperator(BoundAssignmentOperator node, bo
                     rewrittenRight);
             }
 
-            // We have already lowered each argument, but we may need some additional rewriting for the arguments,
```
Is this comment no longer valid?

chsienki commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectOrCollectionInitializerExpression.cs:
```diff
@@ -166,11 +166,6 @@ private BoundExpression MakeDynamicCollectionInitializer(BoundExpression rewritt
-            var rewrittenArguments = VisitList(initializer.Arguments);
-            var rewrittenType = VisitType(initializer.Type);
-
-            // We have already lowered each argument, but we may need some additional rewriting for the arguments,
```
Comment?

chsienki commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs:
```diff
@@ -131,15 +128,15 @@ private BoundExpression VisitIndexerAccess(BoundIndexerAccess node, bool isLeftO
 
                 // We have already lowered each argument, but we may need some additional rewriting for the arguments,
                 // such as generating a params array, re-ordering arguments based on argsToParamsOpt map, inserting arguments for optional parameters, etc.
```
This isn't true now right? We havn't actually lowered the args yet.

333fred replied:
Right. I thought I got all of these, but looks like I missed one.

chsienki commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -391,14 +398,29 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
                 argumentRefKindsOpt = GetEffectiveArgumentRefKinds(argumentRefKindsOpt, parameters);
 
+                ImmutableArray<BoundExpression> rewrittenArguments = arguments.SelectAsArray(
+                    static (argument, context) => argument is BoundDiscardExpression discard
```
Consider expanding this into a loop, to improve readability.

333fred replied:
I actually don't think the expanded version is more readable. It's more than a foreach loop: I will also need to check the arguments length for the 0 case, allocate a builder, and manage it. This is much more concise.

chsienki commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -451,35 +473,111 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+            var actualArguments = new BoundExpression[parameters.Length]; // The actual arguments that will be passed; one actual argument per formal parameter.
+            var storesToTemps = ArrayBuilder<BoundExpression>.GetInstance(arguments.Length);
+            var refKinds = ArrayBuilder<RefKind>.GetInstance(parameters.Length, RefKind.None);
+            var requiresInstanceReceiver = methodOrIndexer.RequiresInstanceReceiver() && methodOrIndexer is not MethodSymbol { MethodKind: MethodKind.Constructor };
```
>methodOrIndexer.RequiresInstanceReceiver() && methodOrIndexer is not MethodSymbol { MethodKind: MethodKind.Constructor }; [](start = 43, length = 121)

We have this in a few places. Consider extracting another externsion method.

333fred replied:
The only two places are in this method. I'm going to leave it as is.

chsienki commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -500,6 +597,147 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+
+                    foreach (var placeholder in interpolationData.ArgumentPlaceholders)
+                    {
+                        if (HasPlaceholderReplacement(placeholder))
```
Consider either removing the assert in `RemovePlaceholderReplacement` or add another method like `RemovePlaceholderReplacementIfPresent(...)`

chsienki commented:
Done with review pass (Iteration 1). 

I havn't looked at the tests particularly closely yet. Had a couple of comments about the binder changes; the lowering changes seem ok, but obviously @AlekseyTs will be able to assert that far better than me :)

333fred commented:
@chsienki addressed your feedback.

---
In reply to: [866455410](https://github.com/dotnet/roslyn/pull/54276#issuecomment-866455410) [](http://example.com/codeflow?ancestors=866455410,866455410)

AlekseyTs commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -6681,4 +6681,17 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
+    <comment>InterpolatedStringHandlerArgumentAttribute is a type name and should not be translated.</comment>
+  </data>
+  <data name="ERR_InterpolatedStringHandlerArgumentLocatedAfterInterpolatedString" xml:space="preserve">
+    <value>Parameter '{0}' is an argument to the interpolated string handler conversion on parameter '{1}', but is specified after the interpolated string constant. Reorder the arguments to move '{0}' before '{1}'.</value>
```
>constant [](http://example.com/codeflow?start=155&length=8)

Is "constant" the right term?

AlekseyTs replied:
>but is specified [](http://example.com/codeflow?start=108&length=16)

"but the corresponding argument is specified"?

AlekseyTs commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -6681,4 +6681,17 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
+    <value>Parameter '{0}' is an argument to the interpolated string handler conversion on parameter '{1}', but is specified after the interpolated string constant. Reorder the arguments to move '{0}' before '{1}'.</value>
+  </data>
+  <data name="ERR_InterpolatedStringHandlerArgumentOptionalNotSpecified" xml:space="preserve">
+    <value>Parameter '{0}' is optional, but is used as an argument to the interpolated string handler conversion on parameter '{1}'. Specify the value of '{0}' before '{1}'.</value>
```
>Parameter '{0}' is optional [](http://example.com/codeflow?start=11&length=27)

Is this about scenario when the corresponding argument is not supplied explicitly? If so, consider clarifying the fact. I assume as long as the argument is there whether the parameter is optional or not is not important. Also, what about ```params``` parameters? they are not optional but can be omitted as well.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -427,5 +529,178 @@ private ImmutableArray<BoundExpression> BindInterpolatedStringParts(BoundUnconve
             return (builderAppendCalls.ToImmutableAndFree(), builderPatternExpectsBool ?? false);
         }
+
+        private BoundExpression BindInterpolatedStringHandlerInMemberCall(
```
>BindInterpolatedStringHandlerInMemberCall [](http://example.com/codeflow?start=32&length=41)

BindInterpolatedStringHandlerArgument?

333fred replied:
It's not specific to whether the parameter has the attribute or not, so I prefer the current name.

AlekseyTs replied:
> It's not specific to whether the parameter has the attribute or not, so I prefer the current name.

I don't think I mentioned attribute in this comment.

333fred replied:
> I don't think I mentioned attribute in this comment.

I ended up inferring it, because `InterpolatedStringHandlerArgumentAttribute` is the name of the attribute.

AlekseyTs replied:
I was referring to method call or indexer access, or quite possibly operator argument.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:
```diff
@@ -2011,7 +2012,9 @@ private BoundFunctionPointerInvocation BindFunctionPointerInvocation(SyntaxNode
                 analyzedArguments.Arguments,
-                diagnostics);
+                diagnostics,
+                boundExpression.Type,
```
>boundExpression [](http://example.com/codeflow?start=16&length=15)

Is this the pointer being invoked? I am not sure if we should treet it as a receiver.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -196,8 +196,17 @@ bool tryBindAsHandlerType([NotNullWhen(true)] out BoundInterpolatedString? resul
+            BindingDiagnosticBag diagnostics,
+            bool isHandlerConversion,
+            ImmutableArray<BoundInterpolatedStringArgumentPlaceholder> additionalConstructorArguments = default,
+            ImmutableArray<RefKind> refKinds = default)
```
>refKinds [](http://example.com/codeflow?start=36&length=8)

additionalRefKinds?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -196,8 +196,17 @@ bool tryBindAsHandlerType([NotNullWhen(true)] out BoundInterpolatedString? resul
+            ImmutableArray<BoundInterpolatedStringArgumentPlaceholder> additionalConstructorArguments = default,
+            ImmutableArray<RefKind> refKinds = default)
         {
+            Debug.Assert(additionalConstructorArguments.IsDefault
```
>additionalConstructorArguments [](http://example.com/codeflow?start=25&length=30)

It feels like it would be much simpler to disallow passing ```default```  for the last two parameters. If it is convenient to allow that at the call sites, after asserting we can reset values using NullToEmpty helper and simplify the logic below by not worrying about ```default``` arrays.

333fred replied:
I think it's important for the call sites to enforce this invariant, to make sure that any future locations where need to bind these does the "right" thing, but I agree we can do `NullToEmpty` after the invariants are verified.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -271,26 +280,119 @@ private BoundInterpolatedString BindUnconvertedInterpolatedStringToHandlerType(B
                 unconvertedInterpolatedString.Type,
                 unconvertedInterpolatedString.HasErrors);
+
+            static void populateArguments(BoundUnconvertedInterpolatedString unconvertedInterpolatedString, ImmutableArray<BoundInterpolatedStringArgumentPlaceholder> additionalConstructorArguments, int baseStringLength, int numFormatHoles, NamedTypeSymbol intType, ArrayBuilder<BoundExpression> argumentsBuilder)
```
>unconvertedInterpolatedString [](http://example.com/codeflow?start=77&length=29)

We only need syntax from this node, let's pass syntax instead.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -271,26 +280,119 @@ private BoundInterpolatedString BindUnconvertedInterpolatedStringToHandlerType(B
+            var nonOutConstructorDiagnostics = BindingDiagnosticBag.GetInstance(template: diagnostics);
+            BoundExpression constructorCall;
+            BoundExpression nonOutConstructorCall = MakeConstructorInvocation(interpolatedStringHandlerType, argumentsBuilder, refKindsBuilder, unconvertedInterpolatedString.Syntax, nonOutConstructorDiagnostics);
+            if (nonOutConstructorCall is not BoundObjectCreationExpression { ResultKind: LookupResultKind.Viable })
```
>nonOutConstructorCall is not BoundObjectCreationExpression { ResultKind: LookupResultKind.Viable } [](http://example.com/codeflow?start=16&length=98)

This approach feels somewhat fragile because we have no idea what went wrong and whether we would prefer reporting the errors rather than trying to bind with an additional out argument. For example, what if the only reason the attempt failed is a missing assembly reference? Should we instead ask to add the reference, or move to the second form? I assume there could be other similar situations. Perhaps we should pay attention to the reasons why “best” among “bad” candidates were rejected. I think we mark candidates with the failure kind during overload resolution.

AlekseyTs replied:
I think it should be fine to follow up on the error detection/recovery separately. A PROTOTYPE comment should be good enough for this PR.

333fred replied:
Filed [https://github.com/dotnet/roslyn/issues/54396](https://github.com/dotnet/roslyn/issues/54396) to follow up.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -271,26 +280,119 @@ private BoundInterpolatedString BindUnconvertedInterpolatedStringToHandlerType(B
+                    // for a different reason, that different reason is the one the user will want to know about. If both or neither failed
+                    // because of this error, we'll report everything.
+
+                    var nonOutConstructorHasArityError = nonOutConstructorDiagnostics.DiagnosticBag?.AsEnumerableWithoutResolution().Any(d => (ErrorCode)d.Code == ErrorCode.ERR_BadCtorArgCount) ?? false;
```
>var nonOutConstructorHasArityError = nonOutConstructorDiagnostics.DiagnosticBag?.AsEnumerableWithoutResolution().Any(d => (ErrorCode)d.Code == ErrorCode.ERR_BadCtorArgCount) ?? false; [](http://example.com/codeflow?start=20&length=183)

I think I agree with Chris, this feels fragile. We should probably do the analysis based on the failure kind for candidates instead, and use the same logic to deside which ones are the "best" among "bad" as we use to decide what errors to report when overload resolution failed. It probably should be logically similar to performing overload resolution on both sets combined at the same time only some candidates allow omitted out argument.

AlekseyTs replied:
>nonOutConstructorDiagnostics.DiagnosticBag? [](http://example.com/codeflow?start=57&length=43)

It appears that the logic can be affected based on the fact whether we are collecting diagnostics or not. That should be avoided. If we absolutely have to pick at the diagnostics, then we should always collect it.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -427,5 +529,178 @@ private ImmutableArray<BoundExpression> BindInterpolatedStringParts(BoundUnconve
+                for (int argumentIndex = 0; argumentIndex < arguments.Count; argumentIndex++)
+                {
+                    // The index in the original parameter list we're looking to match up.
+                    var argumentParameterIndex = memberAnalysisResult.ParameterFromArgument(argumentIndex);
```
>var [](http://example.com/codeflow?start=20&length=3)

int? Any advantage we get from ```var``` here?

333fred replied:
Muscle memory.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -427,5 +529,178 @@ private ImmutableArray<BoundExpression> BindInterpolatedStringParts(BoundUnconve
+            for (int i = 0; i < handlerArgumentIndexes.Length; i++)
+            {
+                int argumentIndex = handlerArgumentIndexes[i];
+                Debug.Assert(argumentIndex != interpolatedStringArgNum);
```
>Debug.Assert(argumentIndex != interpolatedStringArgNum); [](http://example.com/codeflow?start=16&length=56)

Is this validated when we crack the attribute?

333fred replied:
Yes: [https://github.com/dotnet/roslyn/blob/features/interpolated-string/src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs#L781](https://github.com/dotnet/roslyn/blob/features/interpolated-string/src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs#L781). We return default and set the error flag for that case.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -427,5 +529,178 @@ private ImmutableArray<BoundExpression> BindInterpolatedStringParts(BoundUnconve
+
+                RefKind refKind;
+                TypeSymbol placeholderType;
+                if (argumentIndex == BoundInterpolatedStringArgumentPlaceholder.InstanceParameter)
```
>BoundInterpolatedStringArgumentPlaceholder.InstanceParameter [](http://example.com/codeflow?start=37&length=60)

It is quite possible I am missing something, but where do we put this value into the ```handlerArgumentIndexes``` for "Args and parameters were reordered via named parameters, or parameters are missing" scenario?

333fred replied:
You're not missing anything, I am though.

333fred replied:
Added a test to cover this.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -427,5 +529,178 @@ private ImmutableArray<BoundExpression> BindInterpolatedStringParts(BoundUnconve
+                    // If it is optional, then they could otherwise not specify the parameter and that's an error
+                    var originalParameterIndex = handlerParameterIndexes[i];
+                    var parameter = GetCorrespondingParameter(ref memberAnalysisResult, parameters, originalParameterIndex);
+                    if (parameter.IsOptional)
```
>parameter.IsOptional [](http://example.com/codeflow?start=24&length=20)

What about ```params``` parameter?

333fred replied:
Added a params test.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -427,5 +529,178 @@ private ImmutableArray<BoundExpression> BindInterpolatedStringParts(BoundUnconve
+                }
+                else
+                {
+                    var parameter = GetCorrespondingParameter(ref memberAnalysisResult, parameters, argumentIndex);
```
>var parameter = GetCorrespondingParameter(ref memberAnalysisResult, parameters, argumentIndex); [](http://example.com/codeflow?start=20&length=95)

Can we symplly get parameter at ```handlerParameterIndexes[i]``` ordinal?

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs:
```diff
@@ -440,6 +442,19 @@ private void RemovePlaceholderReplacement(BoundValuePlaceholderBase placeholder)
+        {
+            Debug.Assert(placeholder is { });
+            Debug.Assert(_placeholderReplacementMapDoNotUseDirectly is { });
+            _ = _placeholderReplacementMapDoNotUseDirectly.Remove(placeholder);
```
>_ =  [](http://example.com/codeflow?start=12&length=4)

Is the assignment to discard necessary?

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_UsingStatement.cs:
```diff
@@ -493,9 +494,11 @@ private BoundExpression MakeCallWithNoExplicitArgument(MethodArgumentInfo method
                 Debug.Assert(method.ParameterRefKinds.IsDefaultOrEmpty);
             }
+
+            Debug.Assert(methodArgumentInfo.Arguments.All(arg => arg is not BoundConversion { ConversionKind: ConversionKind.InterpolatedStringHandler }));
```
>Debug.Assert(methodArgumentInfo.Arguments.All(arg => arg is not BoundConversion { ConversionKind: ConversionKind.InterpolatedStringHandler })); [](http://example.com/codeflow?start=12&length=143)

Given the name of the containing method, I think the assert should be different. I.e. we should assert there are no arguments.

333fred replied:
Not quite. There can be an extension method argument, for example, and potentially arguments that are allowed to be omitted. The previous asserts cover these cases, and ensure that we only pass non-explicit parameters. Since there was never an explicit parameter, there should not be any interpolated conversions, so we assert that here.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_UsingStatement.cs:
```diff
@@ -505,7 +508,8 @@ private BoundExpression MakeCallWithNoExplicitArgument(MethodArgumentInfo method
                 resultKind: LookupResultKind.Viable,
-                type: method.ReturnType);
+                type: method.ReturnType,
+                argumentsAreVisited: true);
```
>true [](http://example.com/codeflow?start=37&length=4)

Not sure why ```true``` is the right value here.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -156,33 +153,33 @@ public override BoundNode VisitCall(BoundCall node)
             TypeSymbol type,
-            BoundCall? nodeOpt = null)
+            BoundCall? nodeOpt = null,
+            bool argumentsAreVisited = false)
```
>argumentsAreVisited [](http://example.com/codeflow?start=17&length=19)

Not sure what this parameter exactly means.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -366,21 +363,31 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
             ImmutableArray<int> argsToParamsOpt,
             ref ImmutableArray<RefKind> argumentRefKindsOpt,
             out ImmutableArray<LocalSymbol> temps,
+            [NotNullIfNotNull("visitedReceiver")] ref BoundExpression? visitedReceiver,
```
>visitedReceiver [](http://example.com/codeflow?start=71&length=15)

rewrittenReceiver?

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -366,21 +363,31 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+            ThreeState enableCallerInfo = ThreeState.Unknown,
+            bool argumentsWillBeReused = false,
+            ArrayBuilder<BoundExpression>? incomingStores = null,
+            bool argumentsAreVisited = false)
```
>argumentsAreVisited [](http://example.com/codeflow?start=17&length=19)

argumentsAlreadyRewritten?

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -451,35 +473,111 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+                    Debug.Assert(expression is BoundSequence or BoundAssignmentOperator);
+                    if (expression is BoundAssignmentOperator assignment)
+                    {
+                        temporariesBuilder.Add(((BoundLocal)assignment.Left).LocalSymbol);
```
>temporariesBuilder.Add(((BoundLocal)assignment.Left).LocalSymbol); [](http://example.com/codeflow?start=24&length=66)

Why are we not adding temps as we allocate them?

333fred replied:
Will be entirely different with the 2-phase version.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -451,35 +473,111 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+#pragma warning disable CS8619 // https://github.com/dotnet/roslyn/issues/54153
+                    ref BoundExpression resultLocation = ref interpolatedStringHandlerArgumentIndex > -1 && requiresInstanceReceiver
+                        ? ref visitedReceiver!
+                        : ref actualArguments[0];
```
>actualArguments[0]; [](http://example.com/codeflow?start=30&length=19)

When are we using the first argument as the receiver?

333fred replied:
Not really sure I understand the question?

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -451,35 +473,111 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+                        : ref actualArguments[0];
+#pragma warning restore CS8619
+
+                    resultLocation = _factory.Sequence(
```
>resultLocation = _factory.Sequence( [](http://example.com/codeflow?start=20&length=35)

Are we testing scenarios when the receiver is a structure:
- an LValue
- an RValue

Wineed to make sure the right value is used for the receiver and across all iterpolations in the arguments

333fred replied:
I believe I've added tests covering these scenarios.

AlekseyTs replied:
> I believe I've added tests covering these scenarios.

I'd like us to follow up offline to make sure we are on the same page about what I think we should test.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -500,6 +597,144 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+            void visitInterpolatedStringArguments(ref BoundExpression? visitedReceiver, ImmutableArray<RefKind> argumentRefKindsOpt)
+            {
+                // There are handler conversions, so we need to do more complicated reordering of things. We need to
+                // put the receiver into a temp, and then store all the arguments up to the conversion into their own
```
>and then store all the arguments [](http://example.com/codeflow?start=49&length=32)

Why all vs. just those used for handlers?

333fred replied:
Say we have the following call:

```cs
M(Arg1, Arg2, Arg3, Arg4, $"");
```

and that `Arg1` and `Arg3` are used in the string. We need to store `Arg1`, `Arg2`, and `Arg3`, in that order, to preserve lexical ordering. This original version would have also stored `Arg4` because detangling the stores in the single-pass version was too complex, but the new version I will push shortly will skip storing that 4th argument.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -500,6 +597,144 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+                        // only need to register one placeholder per argument
+                        if (HasPlaceholderReplacement(placeholder))
+                        {
+                            continue;
```
>continue [](http://example.com/codeflow?start=28&length=8)

Is this code reachable? It looks like we are allocating distinct instances for each parameter, even if they are going to refer to the same argument. What am I missing?

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -500,6 +597,144 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+                if (requiresInstanceReceiver)
+                {
+                    Debug.Assert(visitedReceiver is not null);
+                    visitedReceiver = _factory.StoreToTemp(visitedReceiver, out var receiverStore, refKind: visitedReceiver.GetRefKind());
```
>visitedReceiver = _factory.StoreToTemp(visitedReceiver, out var receiverStore, refKind: visitedReceiver.GetRefKind()); [](http://example.com/codeflow?start=20&length=118)

Is this going to do the right thing for an LValue struct?

333fred replied:
I'll add a test, but it appears to work correctly from said test (temp is a ref).

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -500,6 +597,144 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+                        ImmutableArray<LocalSymbol>.Empty,
+                        handlerExpressions.ToImmutableAndFree(),
+                        result);
+                    storesToTemps.Add(handlerSequence);
```
>handlerSequence [](http://example.com/codeflow?start=38&length=15)

I am not sure why do we need to add this sequence to storesToTemps, especially with ```result``` in it.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -451,35 +473,111 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+            if (interpolatedStringHandlerArgumentIndex < 0)
+            {
+                // No handler conversions, we can just store all directly to temps
+                VisitAndStoreToTemps(
```
>VisitAndStoreToTemps [](http://example.com/codeflow?start=16&length=20)

If I understand correctly, we are conflating creation of temps needed to reorder named arguments with temps needed to create handlers. I think it would be simple to separate these two concerns. First, we do rewrite for handlers and patch the arguments, then let existing logic to kick in to handle reordering and let it deal with its own temp stores etc.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -624,25 +868,24 @@ static void fillRefKindsBuilder(ImmutableArray<RefKind> argumentRefKindsOpt, Imm
 
         // temporariesBuilder will be null when factory is null.
         private static bool CanSkipRewriting(
```
>CanSkipRewriting [](http://example.com/codeflow?start=28&length=16)

I think we shouldn't change this function and it should only deal with reordering of arguments. The handler rewrite should be a separate pass through the arguments with its own stores, etc.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -673,28 +939,37 @@ private static ImmutableArray<RefKind> GetRefKindsOrNull(ArrayBuilder<RefKind> r
 
         // This fills in the arguments, refKinds and storesToTemps arrays.
-        private void BuildStoresToTemps(
+        private void VisitAndStoreToTemps(
```
>VisitAndStoreToTemps [](http://example.com/codeflow?start=21&length=20)

Conflating visiting and storing to temps feels over complicated, I think these cocerns should be separated.

333fred replied:
> I think these cocerns should be separated

They were separated before this PR. Unfortunately, we need to do these things at the same time for interpolated strings because we need to have those temps around by the time we get to the interpolated string argument. If you have any suggestions on how to structure it differently I'd be happy to hear them, but I don't really have a good structure in mind.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -38,38 +39,67 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
+                }
+                else
+                {
+                    // We're already processing this node, so use the local VisitArguments will have allocated for us
```
>// We're already processing this node, so use the local VisitArguments will have allocated for us [](http://example.com/codeflow?start=20&length=97)

I am having trouble coming up with a scenario that will have this "re-entrancy". Could you elaborate please?

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -38,38 +39,67 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
 
-                var rewrittenArgs = VisitList(appendCall.Arguments);
-                var rewrittenAppendCall = MakeCall(
+                // The append call itself could have and interpolated string conversion that uses the builder local as the receiver
```
>and [](http://example.com/codeflow?start=53&length=3)

Typo "and" vs. "an"?

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -38,38 +39,67 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
+                // The append call itself could have and interpolated string conversion that uses the builder local as the receiver
+                // passed to a nested construction call. This needs to affect the current call to ensure side effects are
+                // preserved, but shouldn't affect any subsequent calls.
+                BoundExpression appendReceiver = builderTemp;
```
>appendReceiver [](http://example.com/codeflow?start=32&length=14)

It doesn't look like we are mutating the value or passing the local ByRef. What exact purpose does this local have?

333fred replied:
Will be used in the next revision.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -260,5 +335,45 @@ public override BoundNode VisitInterpolatedString(BoundInterpolatedString node)
+        {
+            for (int i = 0; i < arguments.Length; i++)
+            {
+                if (arguments[i] is BoundConversion { ConversionKind: ConversionKind.InterpolatedStringHandler, Operand: BoundInterpolatedString operand })
```
>{ ConversionKind: ConversionKind.InterpolatedStringHandler, Operand: BoundInterpolatedString operand } [](http://example.com/codeflow?start=52&length=102)

Should we be checking ExplicitCastInCode here as well.

333fred replied:
This function will be gone in the next revision, but the code is inlined into the new `VisitArguments` function. I've asserted that `ExplicitCastInCode` is never true when the sub condition is true, rather than just checking here, because I want to have an assert verifying that we never hit such a condition and writing an assert in the other direction would be much more complex.

AlekseyTs commented:
Done with review pass (commit 2), tests are not looked at yet.

333fred commented:
I have addressed lowering feedback, but most initial binding feedback is not addressed yet.

---
In reply to: [867888300](https://github.com/dotnet/roslyn/pull/54276#issuecomment-867888300) [](http://example.com/codeflow?ancestors=867888300,867888300)

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_AssignmentOperator.cs:
```diff
@@ -309,20 +311,31 @@ private BoundExpression VisitAssignmentOperator(BoundAssignmentOperator node, bo
-            // such as generating a params array, re-ordering arguments based on argsToParamsOpt map, inserting arguments for optional parameters, etc.
-            ImmutableArray<LocalSymbol> argTemps;
-            rewrittenArguments = MakeArguments(
+            arguments = VisitArguments(
```
>VisitArguments [](http://example.com/codeflow?start=24&length=14)

Was there a bug here that we never lowered aguments?

333fred replied:
No. Arguments were visited earlier, in LocalRewriter_IndexerAccess.cs line 85, but then we held off on calling MakeArguments until this point. However, we now need the output from VisitArguments for the MakeArguments call, so I pushed the visit down to the usage points.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_FunctionPointerInvocation.cs:
```diff
@@ -14,29 +16,36 @@ internal sealed partial class LocalRewriter
         public override BoundNode? VisitFunctionPointerInvocation(BoundFunctionPointerInvocation node)
         {
             var rewrittenExpression = VisitExpression(node.InvokedExpression);
-            var rewrittenArgs = VisitList(node.Arguments);
```
Did we drop rewrite for arguments?

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -38,67 +39,139 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
+                // The append call itself could have an interpolated string conversion that uses the builder local as the receiver
+                // passed to a nested construction call. This needs to affect the current call to ensure side effects are
+                // preserved, but shouldn't affect any subsequent calls.
+                BoundExpression? appendReceiver = builderTemp;
```
>builderTemp [](http://example.com/codeflow?start=50&length=11)

It is still not clear to me what mutation we are trying to preserve here. Could you give a concrete example? If we a trying to prevent mutation of the value stored in ```builderTemp```, perhaps we should allocate a copy temp instead?

333fred replied:
Consider the case of an `AppendFormatted` method that itself has an `InterpolatedStringHandlerArgument("")` attribute on it. The call to `VisitArguments` will replace the receiver with a sequence of saving off the temp and then evaluating the temp again. We don't want the rest of the `AppendFormatted` calls in _this_ interpolated builder to use this new receiver sequence, we want them to continue using the same `BoundLocal` we created above.

333fred replied:
I'm going to add a test to demonstrate this scenario. Look for `InterpolatedStringHandlerArgumentAttribute_AttributeOnAppendFormatCall` in iteration 4 when I push it.

333fred commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -427,5 +528,194 @@ private ImmutableArray<BoundExpression> BindInterpolatedStringParts(BoundUnconve
+            {
+                // Args and parameters were reordered via named parameters, or parameters are missing. Find the correct argument index for each parameter.
+                var handlerArgumentIndexesBuilder = ArrayBuilder<int>.GetInstance(handlerParameterIndexes.Length, fillWithValue: BoundInterpolatedStringArgumentPlaceholder.UnspecifiedParameter);
+                for (int handlerParameterIndex = 0; handlerParameterIndex < handlerParameterIndexes.Length; handlerParameterIndex++)
```
Iteration 4 swaps the order of these loops so I can set the InstanceParameter when we see it, but the content is otherwise the same.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -364,11 +376,219 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+            Symbol methodOrIndexer,
+            ImmutableArray<int> argsToParamsOpt,
+            ref ImmutableArray<RefKind> argumentRefKindsOpt,
+            [NotNullIfNotNull("visitedReceiver")] ref BoundExpression? rewrittenReceiver,
```
>"visitedReceiver" [](http://example.com/codeflow?start=30&length=17)

What does this refer to?

333fred replied:
We really need to do `nameof(parameter)` already.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -364,11 +376,219 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+                Debug.Assert(temps.Count > 0);
+                Debug.Assert(sideEffectsBuilder!.Count > 0);
+                Debug.Assert(!positionsAssignedToTemp.IsNull);
+                Debug.Assert(receiverStored == requiresInstanceReceiver);
```
>receiverStored == requiresInstanceReceiver [](http://example.com/codeflow?start=29&length=42)

This looks suspicious. We need to store the recevier and any other argument only if it used to create a builder.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -364,11 +376,219 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+                else
+                {
+                    visitedArgumentsBuilder[0] = sideEffectsSequence;
+                    positionsAssignedToTemp[0] = false;
```
>positionsAssignedToTemp[0] = false; [](http://example.com/codeflow?start=20&length=35)

Is there a guarantee that the first argument is never cached in a temp?

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -364,11 +376,219 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+
+            for (int i = 0; i < arguments.Length; i++)
+            {
+                InterpolatedStringHandlerData? nullableInterpolatedData = setupInterpolationDataIfNeeded(
```
>nullableInterpolatedData [](http://example.com/codeflow?start=47&length=24)

It doesn't look like we need the entire ```InterpolatedStringHandlerData``` back. An array of placeholders to remove would be sufficient, in my opinion.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -364,11 +376,219 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+
+                    if (creation.Arguments.Length > (interpolationData.HasTrailingHandlerValidityParameter ? 3 : 2))
+                    {
+                        Debug.Assert(!((BoundConversion)argument).ExplicitCastInCode);
```
>Debug.Assert(!((BoundConversion)argument).ExplicitCastInCode); [](http://example.com/codeflow?start=24&length=62)

Is explicit cast fine when condition of the enclosing ```if``` isn't met?

333fred replied:
Yes, because you could explicitly do `(InterpolatedStringHandlerType)$""`, which will be an explicit cast in code with either 2 or 3 arguments.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -364,11 +376,219 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+                        Debug.Assert(sideEffectsBuilder != null);
+                        Debug.Assert(!positionsAssignedTemps.IsNull);
+
+                        // Find the last argument needed by the interpolation, and store from the last stored index all the way up to that argument
```
>Find the last argument needed by the interpolation, and store from the last stored index all the way up to that argument [](http://example.com/codeflow?start=27&length=120)

I don't think this is necessary. I believe we need to store only the arguments that are involved in creation of builders. Those that aren't involved, shouldn't be stored.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -364,11 +376,219 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+            }
+
+            var visitedArgumentsBuilder = ArrayBuilder<BoundExpression>.GetInstance(arguments.Length);
+            ArrayBuilder<BoundExpression>? sideEffectsBuilder = null;
```
>sideEffectsBuilder [](http://example.com/codeflow?start=43&length=18)

I don't think we really need to track any sideffects across all arguments. Each argument  and receiver can be captured individually and can keep the store in a sequence in place.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -364,11 +376,219 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+                var sideEffectsResult = requiresInstanceReceiver ? rewrittenReceiver : visitedArgumentsBuilder[0];
+                Debug.Assert(sideEffectsResult != null);
+
+                var sideEffectsSequence = _factory.Sequence(
```
>sideEffectsSequence [](http://example.com/codeflow?start=20&length=19)

It looks like this patch up wouldn't be necessary if we were capturing sideffects in place.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -364,11 +376,219 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+                        for (int index = lastIndexStored + 1; index <= lastNeededParameter; index++, lastIndexStored++)
+                        {
+                            var paramIndex = argsToParamsOpt.IsDefault ? index : argsToParamsOpt[index];
+                            RefKind argRefKind = argumentRefKindsOpt.RefKinds(index);
```
>argumentRefKindsOpt.RefKinds(index); [](http://example.com/codeflow?start=49&length=36)

Is this call safe on ``default``` array?

333fred replied:
Yes.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -364,11 +376,219 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+
+                            if (refKindsBuilder != null)
+                            {
+                                refKindsBuilder[index] = GetPatchedRefKind(argRefKind, paramRefKind);
```
>refKindsBuilder[index] = GetPatchedRefKind(argRefKind, paramRefKind) [](http://example.com/codeflow?start=32&length=68)

I am not sure why do we need to do this here. It doesn't look like the value is going to depend on anything that we do in this function.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -364,11 +376,219 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
-        /// It is assumed that each argument has already been lowered, but we may need
-        /// additional rewriting for the arguments, such as generating a params array, re-ordering
-        /// arguments based on <paramref name="argsToParamsOpt"/> map, inserting arguments for optional parameters, etc.
+        /// We will lower each argument and we may need additional rewriting for the arguments,
```
>/// We will lower each argument and we may need additional rewriting for the arguments, [](http://example.com/codeflow?start=8&length=87)

Should the comment be restored?

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -390,13 +611,12 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
-            ArrayBuilder<LocalSymbol> temporariesBuilder = ArrayBuilder<LocalSymbol>.GetInstance();
-            rewrittenArguments = _factory.MakeTempsForDiscardArguments(rewrittenArguments, temporariesBuilder);
+            temps ??= ArrayBuilder<LocalSymbol>.GetInstance();
+            rewrittenArguments = _factory.MakeTempsForDiscardArguments(rewrittenArguments, temps);
```
>_factory.MakeTempsForDiscardArguments(rewrittenArguments, temps); [](http://example.com/codeflow?start=33&length=65)

Should this be done during loering of arguments? Presumable we need this argument to create a builder?

333fred replied:
Yeah, you're right. Unlikely scenario, certainly, but we should do this earlier.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -980,11 +1204,12 @@ private static BoundExpression MakeLiteral(SyntaxNode syntax, ConstantValue cons
+                    foreach (BoundExpression s in storesToTemps)
                     {
-                        if (s != null)
+                        Debug.Assert(s is null or BoundAssignmentOperator);
```
>s is null  [](http://example.com/codeflow?start=37&length=10)

How did we end up with nulls in the array?

333fred replied:
`MergeArgumentsAndSideEffects` sets the position to `null` when it eliminates a store.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -980,11 +1204,12 @@ private static BoundExpression MakeLiteral(SyntaxNode syntax, ConstantValue cons
                 if (tempsNeeded > 0)
                 {
-                    foreach (BoundAssignmentOperator s in storesToTemps)
+                    foreach (BoundExpression s in storesToTemps)
```
>BoundExpression [](http://example.com/codeflow?start=29&length=15)

The change looks unnecessary.

AlekseyTs commented:
It looks like there are legitimate test failures

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -427,5 +528,194 @@ private ImmutableArray<BoundExpression> BindInterpolatedStringParts(BoundUnconve
+
+                    if (handlerParameter == BoundInterpolatedStringArgumentPlaceholder.InstanceParameter)
+                    {
+                        handlerArgumentIndexesBuilder[handlerParameterIndex] = handlerParameter;
```
>handlerArgumentIndexesBuilder[handlerParameterIndex] = handlerParameter; [](http://example.com/codeflow?start=24&length=72)

What unit-test covers this fix?

333fred replied:
`InterpolatedStringHandlerArgumentAttribute_ParametersReordered` threw before this fix.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -427,5 +528,194 @@ private ImmutableArray<BoundExpression> BindInterpolatedStringParts(BoundUnconve
+                            // If it is optional, then they could otherwise not specify the parameter and that's an error
+                            var originalParameterIndex = handlerParameterIndexes[i];
+                            var parameter = parameters[originalParameterIndex];
+                            if (parameter.IsOptional || parameter.IsParams)
```
>parameter.IsParams [](http://example.com/codeflow?start=56&length=18)

I believe overload resolution uses more accurarte check for ```params``` parameters

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -377,7 +597,8 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
             ref ImmutableArray<RefKind> argumentRefKindsOpt,
-            out ImmutableArray<LocalSymbol> temps,
+            [NotNull] ref ArrayBuilder<LocalSymbol>? temps,
+            BitVector positionsAssignedToTemp,
```
>positionsAssignedToTemp [](http://example.com/codeflow?start=22&length=23)

Does this really make an observable difference?

333fred replied:
It should have been preventing duplicate stores, but in the next revision it will also be necessary to prevent stores from being optimized away when they shouldn't be.

333fred replied:
Actually, after thinking about this more, I think we can get rid of this and decouple `Visit` and `Make` a bit more.

AlekseyTs commented:
Done with review pass (commit 5), tests are not looked at yet.

333fred commented:
Doesn't seem like it? All the tests passed.

---
In reply to: [869747130](https://github.com/dotnet/roslyn/pull/54276#issuecomment-869747130) [](http://example.com/codeflow?ancestors=869747130,869747130)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:
```diff
@@ -1025,12 +1025,15 @@ public static bool IsValidParams(Symbol member)
 
-            // Note: we need to confirm the "arrayness" on the original definition because
-            // it's possible that the type becomes an array as a result of substitution.
             ParameterSymbol final = member.GetParameters().Last();
```
>member.GetParameters().Last() [](http://example.com/codeflow?start=36&length=29)

I think this important part should be part of ```IsValidParamsParameter```, i.e. only the last parameter can be a valid params parameter.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -364,6 +372,147 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+                {
+                    temps ??= ArrayBuilder<LocalSymbol>.GetInstance();
+                    visitedArgumentsBuilder.Add(_factory.MakeTempForDiscard(discard, temps));
+                    continue;
```
>continue; [](http://example.com/codeflow?start=20&length=9)

It feels like we should be marking this argument as capured in a temp otherwise we are probably capturing it again if it is used by a builder. Also, we should make sure we have struct tests that we discussed offline around discard arguments

333fred replied:
> struct tests that we discussed offline around discard arguments

I'm not sure how you'd observe the results with a discard argument unless I was to use `Unsafe.SkipInit`. Whatever result we set in the builder constructor isn't observable in the body of the method, because the method itself must initialize the struct.

AlekseyTs replied:
VB, for example, doesn't respect out, it treats them as ref and allows reading them

AlekseyTs replied:
VB allows placing the <Out> attribute on the ByRef parameter what C# will see as out

333fred replied:
Ah, I see. I'll add a test like this.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CompoundAssignmentOperator.cs:
```diff
@@ -294,13 +294,41 @@ private BoundIndexerAccess TransformIndexerAccess(BoundIndexerAccess indexerAcce
+                argumentRefKinds,
+                ref transformedReceiver!,
+                out ArrayBuilder<LocalSymbol>? argumentTemps,
+                out bool receiverIsSideEffectSequence);
```
>out bool receiverIsSideEffectSequence [](http://example.com/codeflow?start=16&length=37)

It feels like the unwrapping would apply to any ```BoundSequence``` representing the ```transformedReceiver```. We wouldn't want to evaluate sideffects twice. Given that, I don't think we need this parameter. We should always deconstruct a sequence, Sideeffects go into stores, Locals go into ```temps```.

AlekseyTs commented:
Done with review pass (commit 6)

333fred commented:
@AlekseyTs addressed feedback

---
In reply to: [870176727](https://github.com/dotnet/roslyn/pull/54276#issuecomment-870176727) [](http://example.com/codeflow?ancestors=870176727,870176727)

AlekseyTs reviewed (approved):
LGTM (commit 7), assuming CI is passing.

jcouv commented:
> ");

Do we have a test for an API with two custom handlers, each with a handler argument attribute?
It may be interesting if the second handler argument attribute refers to the first custom handler parameter. 
Chaining example: `void M(int i1, [Argument("i1")] CustomHandler1 c1, [Argument("i1", "c1"] CustomHandler2 c2)`

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:10268 in d57ccfe. [](commit_id = d57ccfe9bb491cb31f0a04ed773b36e4fa2af34f, deletion_comment = False)

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -3019,8 +3022,26 @@ private BoundExpression BindArgumentExpression(BindingDiagnosticBag diagnostics,
+
+            void reportUnsafeIfNeeded(MemberResolutionResult<TMember> methodResult, BindingDiagnosticBag diagnostics, BoundExpression argument, TypeWithAnnotations parameterTypeWithAnnotations)
+            {
+                // NOTE: for some reason, dev10 doesn't report this for indexer accesses.
```
Do we need to maintain such legacy behavior for new scenario? I understand that `if (!kind.IsIdentity)` case needs to stay like this, but the `if (kind.IsInterpolatedStringHandler)` is not bound by prior behavior.

333fred replied:
We could potentially do this, but I'd rather just keep it simple and have one set of unsafe binding rules for everything.

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -427,5 +528,194 @@ private ImmutableArray<BoundExpression> BindInterpolatedStringParts(BoundUnconve
+                    hasErrors: true);
+            }
+
+            var handlerParameterIndexes = interpolatedStringParameter.InterpolatedStringHandlerArgumentIndexes;
```
nit: not related to this PR, consider updating the xml doc on `InterpolatedStringHandlerArgumentIndexes` with references to the special values such as `BoundInterpolatedStringArgumentPlaceholder.UnspecifiedParameter` and others

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -271,26 +283,115 @@ private BoundInterpolatedString BindUnconvertedInterpolatedStringToHandlerType(B
+
+            var nonOutConstructorDiagnostics = BindingDiagnosticBag.GetInstance(withDiagnostics: true, withDependencies: diagnostics.AccumulatesDependencies);
+            BoundExpression constructorCall;
+            BoundExpression nonOutConstructorCall = MakeConstructorInvocation(interpolatedStringHandlerType, argumentsBuilder, refKindsBuilder, unconvertedInterpolatedString.Syntax, nonOutConstructorDiagnostics);
```
If I'm reading the spec right, I'd expect we try to find  the constructor with final `out bool` parameter first, then retry without it.
[https://github.com/dotnet/csharplang/blob/main/proposals/improved-interpolated-strings.md#constructor-resolution](https://github.com/dotnet/csharplang/blob/main/proposals/improved-interpolated-strings.md#constructor-resolution)

333fred replied:
I need to reverse the order, thanks for pointing it out.

333fred replied:
Actually, thinking about it more I had a reason for doing that in the order I put it in the spec. I'll reverse this.

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -271,26 +283,115 @@ private BoundInterpolatedString BindUnconvertedInterpolatedStringToHandlerType(B
+            if (nonOutConstructorCall is not BoundObjectCreationExpression { ResultKind: LookupResultKind.Viable })
+            {
+                // MakeConstructorInvocation can call CoerceArguments on the builder if overload resolution succeeded ignoring accessibility, which
+                // could still end up not succeeding, and that would end up changing the arguments. So we want to clear and repopulate.
```
Thanks for the comment. Was wondering about that

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -271,26 +283,115 @@ private BoundInterpolatedString BindUnconvertedInterpolatedStringToHandlerType(B
+                            constructorCall = nonOutConstructorCall;
+                            diagnostics.AddRangeAndFree(nonOutConstructorDiagnostics);
+                            diagnostics.AddRangeAndFree(outConstructorDiagnostics);
+                            break;
```
nit: extra blank line below

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -271,26 +283,115 @@ private BoundInterpolatedString BindUnconvertedInterpolatedStringToHandlerType(B
+            var nonOutConstructorDiagnostics = BindingDiagnosticBag.GetInstance(withDiagnostics: true, withDependencies: diagnostics.AccumulatesDependencies);
+            BoundExpression constructorCall;
+            BoundExpression nonOutConstructorCall = MakeConstructorInvocation(interpolatedStringHandlerType, argumentsBuilder, refKindsBuilder, unconvertedInterpolatedString.Syntax, nonOutConstructorDiagnostics);
+            if (nonOutConstructorCall is not BoundObjectCreationExpression { ResultKind: LookupResultKind.Viable })
```
nit: consider inverting `if` by removing `not` from pattern

333fred replied:
I think it reads better as is.

jcouv commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -2025,6 +2025,9 @@
   <Node Name="BoundInterpolatedStringArgumentPlaceholder" Base="BoundValuePlaceholderBase">
     <Field Name="Type" Type="TypeSymbol" Override="true" Null="disallow" />
+    <!-- The index in the containing member of the argument this is the placeholder for. Should be a positive number or
+         one of the constants in the other part of the partial class. -->
```
I think `BoundInterpolatedStringArgumentPlaceholder` doesn't survive lowering. If so, we can add a comment (see `BoundDeconstructValuePlaceholder` for example) and enforce it in `LocalRewritingValidator`.

jcouv commented on src/Compilers/CSharp/Portable/BoundTree/InterpolatedStringHandlerData.cs:
```diff
@@ -10,21 +11,38 @@ namespace Microsoft.CodeAnalysis.CSharp
         {
             Debug.Assert(construction is BoundObjectCreationExpression or BoundDynamicObjectCreationExpression or BoundBadExpression);
+            Debug.Assert(!placeholders.IsDefault);
+            // Only the last placeholder should be the out parameter.
```
>should [](http://example.com/codeflow?start=41&length=6)

nit: may be

jcouv commented:
> ");

Although the official `InterpolatedStringHandlerArgumentAttribute` is defined to not allow multiple, we should probably have a test where the definition does allow multiple and we use it multiple times.$

---
In reply to: [871707137](https://github.com/dotnet/roslyn/pull/54276#issuecomment-871707137) [](http://example.com/codeflow?ancestors=871707137)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:10268 in d57ccfe. [](commit_id = d57ccfe9bb491cb31f0a04ed773b36e4fa2af34f, deletion_comment = False)

jcouv commented:
> ");

Do we disallow the argument attribute on `ref` or `out` parameters?
Either way, consider adding a test for `out`. I assume the conversion rules apply, but the arguments attribute plays no role.
```
void M(..., [Argument(...) out CustomHandler result)
{
   result = $"....";
}
```

---
In reply to: [871839205](https://github.com/dotnet/roslyn/pull/54276#issuecomment-871839205) [](http://example.com/codeflow?ancestors=871839205,871707137)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:10268 in d57ccfe. [](commit_id = d57ccfe9bb491cb31f0a04ed773b36e4fa2af34f, deletion_comment = False)

jcouv commented:
> ");

4) Should fall out (since should re-use `SourceComplexParameterSymbol`), but consider adding tests for argument attribute used on local function, lambda and delegate.

---
In reply to: [871840029](https://github.com/dotnet/roslyn/pull/54276#issuecomment-871840029) [](http://example.com/codeflow?ancestors=871840029,871839205,871707137)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:10268 in d57ccfe. [](commit_id = d57ccfe9bb491cb31f0a04ed773b36e4fa2af34f, deletion_comment = False)

jcouv commented:
> ");

5) Consider testing default parameter value `void M(..., [Argument(...)] CustomHandler x = $"...")`. The interpolated string may be a constant. I'm not sure what to expect in such case.

---
In reply to: [871842028](https://github.com/dotnet/roslyn/pull/54276#issuecomment-871842028) [](http://example.com/codeflow?ancestors=871842028,871840029,871839205,871707137)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:10268 in d57ccfe. [](commit_id = d57ccfe9bb491cb31f0a04ed773b36e4fa2af34f, deletion_comment = False)

jcouv commented:
> ");

6) Consider testing dictionary initializer with indexer using custom handler and argument attribute: `new C() { [..., $"..."] = 42 }`

---
In reply to: [871847782](https://github.com/dotnet/roslyn/pull/54276#issuecomment-871847782) [](http://example.com/codeflow?ancestors=871847782,871842028,871840029,871839205,871707137)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:10268 in d57ccfe. [](commit_id = d57ccfe9bb491cb31f0a04ed773b36e4fa2af34f, deletion_comment = False)

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectCreationExpression.cs:
```diff
@@ -35,22 +36,31 @@ public override BoundNode VisitObjectCreationExpression(BoundObjectCreationExpre
+                ref tempsBuilder);
 
             BoundExpression rewrittenObjectCreation;
+            var temps = tempsBuilder?.ToImmutableAndFree() ?? default;
```
I think `tempsBuilder` is not-null after `MakeArguments`.

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -759,6 +900,23 @@ bool isLambdaConversion(BoundExpression expr)
                 => expr is BoundConversion conv && conv.ConversionKind == ConversionKind.AnonymousFunction;
         }
 
+        private static RefKind GetPatchedRefKind(RefKind argRefKind, RefKind paramRefKind)
```
nit: this is only used once (above), could just be a local function?

333fred replied:
I'll just re-inline the code, I was using it in multiple places before.

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -364,6 +371,155 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
             }
         }
 
+        private ImmutableArray<BoundExpression> VisitArguments(
```
Method probably deserves a comment ;-)

jcouv commented on src/Compilers/CSharp/Portable/Lowering/DiagnosticsPass_ExpressionTrees.cs:
```diff
@@ -710,6 +710,13 @@ public override BoundNode VisitConversion(BoundConversion node)
+                case ConversionKind.InterpolatedStringHandler:
+                    if (_inExpressionLambda)
+                    {
+                        Error(ErrorCode.ERR_ExpressionTreeContainsInterpolatedStringHandlerConversion, node);
```
Looks like we don't have a test for this

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -364,6 +371,155 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+                                    RefKind argRefKind = argumentRefKindsOpt.RefKinds(argIndex);
+                                    RefKind paramRefKind = parameters[paramIndex].RefKind;
+                                    var visitedArgument = visitedArgumentsBuilder[argIndex];
+                                    local = _factory.StoreToTemp(visitedArgument, out store, refKind: paramRefKind == RefKind.In ? RefKind.In : argRefKind);
```
Should we avoid creating extra temps if `visitedArgument` is already a temp?
Same question for `visitedReceiver` above (line 477).

Related question: what do we expect in terms of side-effects when using a `ref` argument and an argument attribute?
For instance, when calling `M(ref string x, [ArgumentAttribute("x")] CustomHandler handler)`, will `M` receive a value for `x` that is affected by the handler conversion, or are side-effects from the handler conversion supposed to be discarded?

333fred replied:
I don't think we should be trying to optimize temps in here. If we store an extra temp, so be it.

As to refs, refkinds must match. So if you pass something by ref, it can change values. There are tests covering this behavior.

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -364,6 +371,155 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+
+            ImmutableArray<BoundInterpolatedStringArgumentPlaceholder> setupInterpolationDataIfNeeded(
+                int argumentIndex,
+                ref bool receiverStored,
```
nit: consider "receiverAssignedToTemp" to align with `positionsAssignedToTemp`

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -364,6 +371,155 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+                        foreach (var placeholder in interpolationData.ArgumentPlaceholders)
+                        {
+                            // Replace each needed placeholder with a sequence of store and evaluate the temp.
+                            var argIndex = placeholder.ArgumentIndex;
```
Consider asserting that `argIndex < argumentIndex` (enforcing that `visitedArgumentsBuilder[argIndex]` usages below only reference already visited arguments)

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -364,6 +371,155 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+                                    {
+                                        BoundSequence { Value: BoundLocal l } => l,
+                                        BoundLocal l => l, // Can happen for discard arguments
+                                        var u => throw ExceptionUtilities.UnexpectedValue(u)
```
>u [](http://example.com/codeflow?start=90&length=1)

maybe `UnexpectedValue(u.Kind)`?

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -364,6 +371,155 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+                                    local = visitedArgumentsBuilder[argIndex] switch
+                                    {
+                                        BoundSequence { Value: BoundLocal l } => l,
+                                        BoundLocal l => l, // Can happen for discard arguments
```
nit: I didn't follow how this could happen given the handling in `case >= 0:` below (which always puts a `BoundSequence` with a `BoundLocal` value into the `visitedArgumentsBuilder`). I'm just curious

333fred replied:
When we assign a discard temp, we mark the position as assigned because that local doesn't need a temp.

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -364,6 +371,155 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+                                    var paramIndex = argsToParamsOpt.IsDefault ? argIndex : argsToParamsOpt[argIndex];
+                                    RefKind argRefKind = argumentRefKindsOpt.RefKinds(argIndex);
+                                    RefKind paramRefKind = parameters[paramIndex].RefKind;
+                                    var visitedArgument = visitedArgumentsBuilder[argIndex];
```
nit: I think we can assert that previously visited values are not handler placeholders. Consider adding

333fred replied:
I don't really understand what you're saying here.

jcouv replied:
I'm suggesting `Debug.Assert(visitedArgument  is not BoundInterpolatedStringArgumentPlaceholder)`

333fred replied:
I guess I don't really know what that assert buys us.

333fred replied:
In general, placeholders shouldn't survive a visit. BoundInterpolatedStringArgumentPlaceholders are placeholders like any other.

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -364,6 +371,155 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+                }
+            }
+
+            ImmutableArray<BoundInterpolatedStringArgumentPlaceholder> setupInterpolationDataIfNeeded(
```
nit: `addInterpolationPlaceholderReplacements` to align with `AddPlaceholderReplacement` and `RemovePlaceholderReplacement`?

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -260,5 +330,26 @@ public override BoundNode VisitInterpolatedString(BoundInterpolatedString node)
         }
+
+        [Conditional("DEBUG")]
+        private void AssertNoImplicitInterpolatedStringHandlerConversions(ImmutableArray<BoundExpression> arguments, bool allowConversionsWithNoContext = false)
```
static?

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -260,5 +330,26 @@ public override BoundNode VisitInterpolatedString(BoundInterpolatedString node)
+                {
+                    if (arg is BoundConversion { Conversion: { Kind: ConversionKind.InterpolatedStringHandler }, ExplicitCastInCode: false, Operand: BoundInterpolatedString @string })
+                    {
+                        Debug.Assert(((BoundObjectCreationExpression)@string.InterpolationData!.Value.Construction).Arguments.All(
```
nit: would it be simpler to check `ArgumentPlaceholders` rather than looking at the arguments of `Construction`? Or maybe both?

333fred replied:
I'd rather assert against the thing that is actually going to be emitted.

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -38,67 +39,136 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
+                Debug.Assert(construction.ArgumentRefKindsOpt[^1] == RefKind.Out);
+
+                BoundInterpolatedStringArgumentPlaceholder trailingParameter = data.ArgumentPlaceholders[^1];
+                TypeSymbol localType = trailingParameter.Type!;
```
Can `trailingParameter.Type` ever be null?

333fred replied:
No.

jcouv replied:
Then we don't need a suppression

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -38,67 +39,136 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
+                var outLocal = _factory.SynthesizedLocal(localType);
+                appendShouldProceedLocal = _factory.Local(outLocal);
+
+                AddPlaceholderReplacement(trailingParameter, appendShouldProceedLocal);
```
Every call to `AddPlaceholderReplacement` should have a corresponding call to remove placeholder replacement

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -38,67 +39,136 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
+            var resultExpressions = ArrayBuilder<BoundExpression>.GetInstance(node.Parts.Length + 1);
             foreach (var currentPart in node.Parts)
             {
                 var appendCall = (BoundCall)currentPart;
```
The assertion above (`Debug.Assert(node.Parts.All(static p => p is BoundCall or BoundStringInsert { Value: BoundCall }));` on line 49) probably should be tightened to only expect `BoundCall` parts.

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -38,67 +39,136 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
 
-                Debug.Assert(usesBoolReturn == (rewrittenAppendCall.Type!.SpecialType == SpecialType.System_Boolean));
-                if (usesBoolReturn)
+                Debug.Assert(usesBoolReturn == (appendCall.Type!.SpecialType == SpecialType.System_Boolean));
```
This assertion is duplicate from one above (line 79). I  think it made more sense before (checking `rewrittenAppendCall` instead)

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -38,67 +39,136 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
-            // var builder = new BuilderType(baseStringLength, numFormatHoles);
-            Debug.Assert(data.Construction is BoundObjectCreationExpression);
-            var builderConstruction = _factory.AssignmentExpression(builderTemp, (BoundExpression)VisitObjectCreationExpression((BoundObjectCreationExpression)data.Construction));
+            // var handler = new HandlerType(baseStringLength, numFormatHoles, ...InterpolatedStringHandlerArgumentAttribute parameters, <optional> out bool handlerIsValid);
```
>handlerIsValid [](http://example.com/codeflow?start=157&length=14)

`<optional> out bool appendShouldProceed`?

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -38,67 +39,136 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
+                    appendShouldProceedLocal is not null
+                        ? ImmutableArray.Create(appendShouldProceedLocal.LocalSymbol)
+                        : ImmutableArray<LocalSymbol>.Empty,
+                    resultExpressions.ToImmutableAndClear(),
```
nit: `ImmutableArray.Create(handlerConstructionAssignment)` since this is always exactly one expression?

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -38,67 +39,136 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
+                // appendShouldProceedLocal && sequence
+                var appendAnd = _factory.LogicalAnd(appendShouldProceedLocal, appendCallsSequence);
+
+                BoundExpression result = appendAnd;
```
Unnecessary assignment

jcouv reviewed (commented):
Done reviewing code (iteration 8). Will looks at tests tomorrow

jcouv commented:
>                 Diagnostic(ErrorCode.ERR_TypeIsNotAnInterpolatedStringHandlerType, "InterpolatedStringHandlerArgumentAttribute").WithArguments("string").WithLocation(8, 27)

Not blocking this PR: I don't think the spec details the validation rules for attribute.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:5705 in d57ccfe. [](commit_id = d57ccfe9bb491cb31f0a04ed773b36e4fa2af34f, deletion_comment = False)

jcouv commented:
>             // we just bind to that method without checking the handler attribute.

ni: Do we have/want a similar test where the caller isn't giving a constant?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:5730 in d57ccfe. [](commit_id = d57ccfe9bb491cb31f0a04ed773b36e4fa2af34f, deletion_comment = False)

333fred commented:
>             // we just bind to that method without checking the handler attribute.

That doesn't have any meaningful impact on the scenario.

---
In reply to: [872440668](https://github.com/dotnet/roslyn/pull/54276#issuecomment-872440668) [](http://example.com/codeflow?ancestors=872440668,872440668)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:5730 in d57ccfe. [](commit_id = d57ccfe9bb491cb31f0a04ed773b36e4fa2af34f, deletion_comment = False)

333fred commented:
>                 Diagnostic(ErrorCode.ERR_TypeIsNotAnInterpolatedStringHandlerType, "InterpolatedStringHandlerArgumentAttribute").WithArguments("string").WithLocation(8, 27)

Not sure I understand what you mean.

---
In reply to: [872438673](https://github.com/dotnet/roslyn/pull/54276#issuecomment-872438673) [](http://example.com/codeflow?ancestors=872438673,872438673)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:5705 in d57ccfe. [](commit_id = d57ccfe9bb491cb31f0a04ed773b36e4fa2af34f, deletion_comment = False)

jcouv commented:
> ");

7) I couldn't find tests with argument attribute used in an extension method. We should invoke such extension method as an extension or directly.
8) Not related to this PR: Do we have a test for an error scenario `(CustomHandler)$"..."` where the conversion fails, but would have succeeded if we were in an invocation?
9) `[InterpolatedStringHandlerArgumentAttribute(null)]`

---
In reply to: [871853568](https://github.com/dotnet/roslyn/pull/54276#issuecomment-871853568) [](http://example.com/codeflow?ancestors=871853568,871847782,871842028,871840029,871839205,871707137)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:10268 in d57ccfe. [](commit_id = d57ccfe9bb491cb31f0a04ed773b36e4fa2af34f, deletion_comment = False)

jcouv commented:
>         public void SideEffects(bool useBoolReturns, bool validityParameter)

Could we add a similar test but with `param2` being a `ref` and the handler constructor modifying it?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:8875 in d57ccfe. [](commit_id = d57ccfe9bb491cb31f0a04ed773b36e4fa2af34f, deletion_comment = False)

333fred commented:
>         public void SideEffects(bool useBoolReturns, bool validityParameter)

See `StructReceiver_Lvalue`

---
In reply to: [872462032](https://github.com/dotnet/roslyn/pull/54276#issuecomment-872462032) [](http://example.com/codeflow?ancestors=872462032,872462032)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:8875 in d57ccfe. [](commit_id = d57ccfe9bb491cb31f0a04ed773b36e4fa2af34f, deletion_comment = False)

jcouv commented:
>         public void SideEffects(bool useBoolReturns, bool validityParameter)

Ah, cool. I didn't realize that scenario was going to be an error, so was worried. Thanks!

---
In reply to: [872464569](https://github.com/dotnet/roslyn/pull/54276#issuecomment-872464569) [](http://example.com/codeflow?ancestors=872464569,872462032)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:8875 in d57ccfe. [](commit_id = d57ccfe9bb491cb31f0a04ed773b36e4fa2af34f, deletion_comment = False)

jcouv reviewed (commented):
Done with review pass (iteration 8)

333fred commented:
>         public void SideEffects(bool useBoolReturns, bool validityParameter)

Sorry, I linked the wrong test. I meant `StructParameter_ByRef`. Regardless, the scenario is tested.

---
In reply to: [872470655](https://github.com/dotnet/roslyn/pull/54276#issuecomment-872470655) [](http://example.com/codeflow?ancestors=872470655,872462032)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:8875 in d57ccfe. [](commit_id = d57ccfe9bb491cb31f0a04ed773b36e4fa2af34f, deletion_comment = False)

333fred commented:
> ");

1. I'll add to this pr.
2. I don't think this is important, we generally don't validate these things.
3. We have tests for ref. For an out parameter, you would never be able to pass one of these in the first place because it needs to be a location.
4. I'll add it to the test plan issue for after merge.
5. I'll add it to the test plan issue for after merge, but IMO there's not much interesting here. We don't copy the expression into a new location for default parameters, we copy values. Either that doesn't compile because you  don't have a conversion from string on your builder or it compiles and is a user-defined conversion.
6. I'll add it to the test plan.
7. I'll add it to the test plan.
8. I'll add it to the test plan.
9. See `InterpolatedStringHandlerArgumentAttributeError_NullConstant*`.

---
In reply to: [872460526](https://github.com/dotnet/roslyn/pull/54276#issuecomment-872460526) [](http://example.com/codeflow?ancestors=872460526,871707137)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:10268 in d57ccfe. [](commit_id = d57ccfe9bb491cb31f0a04ed773b36e4fa2af34f, deletion_comment = False)

333fred commented:
> ");

Added to [https://github.com/dotnet/roslyn/issues/51499#issuecomment-872496524](https://github.com/dotnet/roslyn/issues/51499#issuecomment-872496524).

---
In reply to: [872494519](https://github.com/dotnet/roslyn/pull/54276#issuecomment-872494519) [](http://example.com/codeflow?ancestors=872494519,872460526,871707137)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:10268 in d57ccfe. [](commit_id = d57ccfe9bb491cb31f0a04ed773b36e4fa2af34f, deletion_comment = False)

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CompoundAssignmentOperator.cs:
```diff
@@ -294,13 +294,38 @@ private BoundIndexerAccess TransformIndexerAccess(BoundIndexerAccess indexerAcce
+                argsToParamsOpt,
+                argumentRefKinds,
+                ref transformedReceiver!,
+                out ArrayBuilder<LocalSymbol>? argumentTemps);
```
nit: `argumentTemps` can't be null after `VisitArguments`. We don't need the null check below

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_FunctionPointerInvocation.cs:
```diff
@@ -14,29 +14,42 @@ internal sealed partial class LocalRewriter
+                argsToParamsOpt: default,
+                argumentRefKindsOpt: argumentRefKindsOpt,
+                rewrittenReceiver: ref discardedReceiver,
+                temps: out ArrayBuilder<LocalSymbol>? temps);
```
nit: Not nullable

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs:
```diff
@@ -121,38 +118,44 @@ private BoundExpression VisitIndexerAccess(BoundIndexerAccess node, bool isLeftO
+                    argsToParamsOpt,
+                    argumentRefKindsOpt,
+                    ref rewrittenReceiver!,
+                    out ArrayBuilder<LocalSymbol>? temps);
```
nit: same here

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectCreationExpression.cs:
```diff
@@ -35,22 +36,30 @@ public override BoundNode VisitObjectCreationExpression(BoundObjectCreationExpre
+                node.ArgsToParamsOpt,
+                argumentRefKindsOpt,
+                ref receiverDiscard,
+                out ArrayBuilder<LocalSymbol>? tempsBuilder);
```
nit: also here

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectOrCollectionInitializerExpression.cs:
```diff
@@ -183,7 +177,17 @@ private BoundExpression MakeDynamicCollectionInitializer(BoundExpression rewritt
+                initializer.ArgsToParamsOpt,
+                argumentRefKindsOpt,
+                ref rewrittenReceiver,
+                out ArrayBuilder<LocalSymbol>? temps);
```
nit: also here

jcouv reviewed (approved):
LGTM Thanks (iteration 9). Some nullability nits

333fred commented:
I'll take care of nullability nits in the next PR.
